src/parsers/utils/fortran.js

```diff
@@ -0,0 +1,115 @@
+
+        return [key, value];
+    });
+}
+
+/**
+ * @summary Extracts an array of the key value pairs from a Fortran namelist.
+ * @param {String} data
+ * @returns {Object[]}
+ */
+function extractKeyValuePairs(data) {
+    const output = {};
+    const numberPairs = extractPairs(data, regex.fortran.numberKeyValue, Number); // FIXME: fails to convert numbers like 1.234D-567 due to 'D'
+    const stringPairs = extractPairs(data, regex.fortran.stringKeyValue, String);
+    const booleanPairs = extractPairs(data, regex.fortran.booleanKeyValue, Boolean);
+    const numberArrayPairs = extractPairs(data, regex.fortran.numberArrayKeyValue, Number, true);
+    const stringArrayPairs = extractPairs(data, regex.fortran.stringArrayKeyValue, String, true);
+    const booleanArrayPairs = extractPairs(data, regex.fortran.booleanArrayKeyValue, Boolean, true);
+
+    [...numberPairs, ...stringPairs, ...booleanPairs].forEach((pair) => {
+        // eslint-disable-next-line prefer-destructuring
+        output[pair[0]] = pair[1];
+    });
+
+    [numberArrayPairs, stringArrayPairs, booleanArrayPairs].forEach((arrayPairs) => {
+        arrayPairs.forEach(([key, value]) => {
+            const [index, actualValue] = value;
+            if (!output[key]) output[key] = [];
+            output[key][index - 1] = actualValue; // to start arrays from index 0, while fortran lists start from 1
+        });
+    });
+
+    return output;
+}
+
+/**
+ * @summary Extracts namelist data from a string.
+ * @param {String} text
+ * @returns {Object}
+ */
+function extractNamelistData(text) {
+    const namelistNameRegex = /^&(\w+)/gm;
+    const matches = Array.from(text.matchAll(namelistNameRegex));
+    const namelistNames = matches.map((match) => match[1].toLowerCase());
+    const namelists = {};
+
+    namelistNames.forEach((namelistName) => {
+        const _regex = regex.fortran.namelists(namelistName);
+        const data = text.match(_regex)[2];
+
+        namelists[namelistName] = extractKeyValuePairs(data);
+    });
+    return namelists;
+}
+
+/** s
+ * Parses Fortran namelists and cards data from a string.
+ *
+ * @summary Parses Fortran namelists and cards data from a QE input file string.
+ * @param {String} text - The text to parse.
+ * @throws {Error} If no namelist data is found in `text`.
+ * @throws {Error} If no cards data is found in `text`.
+ * @returns {Object} An object containing the parsed namelist and cards data. The exact structure of this object will depend on the structure of the namelist and cards data in `text`.
+ */
+function parseFortranFile(text) {
+    let output = {};
+    try {
+        output = extractNamelistData(text);
+    } catch (err) {
+        throw new Error("Incorrect fortran file");
+    }
+
+    const match = regex.fortran.cards.exec(text);
+    // eslint-disable-next-line prefer-destructuring
+    output.cards = match[0];
+    return output;
+}
+
+export class FortranParser extends BaseParser {
```

Comment:
Should be defined as a mixin per the above

Reply:
The FortranParser is a mixin, not a class itself?

Reply:
To test `FortranParserMixin`'s `parseNamelists()` i've created a `TestClass` extending `BaseParser `with `FortranParserMixin`